src/pair_rann.cpp

```diff
@@ -0,0 +1,1247 @@
+	double dssumy[jnum][net1.maxlayer];
+	double dsz[jnum][net1.maxlayer];
+	double dssumz[jnum][net1.maxlayer];
+	//energy output with forces from analytical derivatives
+	double dsum1;
+	int f = net1.dimensions[0];
+	for (i=0;i<net1.layers-1;i++){
+		for (j=0;j<net1.dimensions[i+1];j++){
+			//energy forward propagation
+			sum[j]=0;
+			for (k=0;k<net1.dimensions[i];k++){
+				if (i==0&&j==0){
+					layer[k]=features[k];
+				}
+				sum[j] += net1.Weights[i][j*net1.dimensions[i]+k]*layer[k];
+			}
+			sum[j] += net1.Biases[i][j];
+			dsum1 = activation[itype][i]->dactivation_function(sum[j]);
+			sum[j] = activation[itype][i]->activation_function(sum[j]);
+			if (i==L-1){
+				energy[j] = sum[j];
+				if (eflag_atom)eatom[i1]=sum[j];
+				if (eflag_global){eng_vdwl +=sum[j];}
+			}
+			//force propagation
+			for (jj=0;jj<jnum;jj++){
+//				if (Bij[jj]==false)continue;
+				dlayersumx[jj][j]=0;
+				dlayersumy[jj][j]=0;
+				dlayersumz[jj][j]=0;
+				dssumx[jj][j]=0;
+				dssumy[jj][j]=0;
+				dssumz[jj][j]=0;
+				for (k=0;k<net1.dimensions[i];k++){
+					if (i==0&&j==0){
+						dlayerx[jj][k]=dfeaturesx[jj*f+k];
+						dlayery[jj][k]=dfeaturesy[jj*f+k];
+						dlayerz[jj][k]=dfeaturesz[jj*f+k];
+						dsx[jj][k]=-sx[jj*f+k];
+						dsy[jj][k]=-sy[jj*f+k];
+						dsz[jj][k]=-sz[jj*f+k];
+					}
+					double w1 = net1.Weights[i][j*net1.dimensions[i]+k];
+					dlayersumx[jj][j] += w1*dlayerx[jj][k];
+					dlayersumy[jj][j] += w1*dlayery[jj][k];
+					dlayersumz[jj][j] += w1*dlayerz[jj][k];
+					dssumx[jj][j] += w1*dsx[jj][k];
+					dssumy[jj][j] += w1*dsy[jj][k];
+					dssumz[jj][j] += w1*dsz[jj][k];
+				}
+				dlayersumx[jj][j] = dsum1*dlayersumx[jj][j];
+				dlayersumy[jj][j] = dsum1*dlayersumy[jj][j];
+				dlayersumz[jj][j] = dsum1*dlayersumz[jj][j];
+				dssumx[jj][j] *= dsum1;
+				dssumy[jj][j] *= dsum1;
+				dssumz[jj][j] *= dsum1;
+				if (i==L-1 && jj < (jnum-1)){
+//					j1 = jlist[jj];
+//					j1 &= NEIGHMASK;
+					j1 = jl[jj];
+					force[j1][0]+=dlayersumx[jj][j];
+					force[j1][1]+=dlayersumy[jj][j];
+					force[j1][2]+=dlayersumz[jj][j];
+					fm[j1][0]+=dssumx[jj][j];
+					fm[j1][1]+=dssumy[jj][j];
+					fm[j1][2]+=dssumz[jj][j];
+				}
+			}
+			if (i==L-1){
+				j1 = ilist[ii];
+				jj = jnum-1;
+				force[j1][0]+=dlayersumx[jj][j];
+				force[j1][1]+=dlayersumy[jj][j];
+				force[j1][2]+=dlayersumz[jj][j];
+				fm[j1][0]+=dssumx[jj][j];
+				fm[j1][1]+=dssumy[jj][j];
+				fm[j1][2]+=dssumz[jj][j];
+			}
+		}
+		//update values for next iteration
+		for (j=0;j<net1.dimensions[i+1];j++){
+			layer[j]=sum[j];
+			for (jj=0;jj<jnum;jj++){
+//				if (Bij[jj]==false)continue;
+				dlayerx[jj][j] = dlayersumx[jj][j];
+				dlayery[jj][j] = dlayersumy[jj][j];
+				dlayerz[jj][j] = dlayersumz[jj][j];
+				dsx[jj][j] = dssumx[jj][j];
+				dsy[jj][j] = dssumy[jj][j];
+				dsz[jj][j] = dssumz[jj][j];
+			}
+		}
+	}
+}
+
+
+
+//treats # as starting a comment to be ignored.
+int PairRANN::count_words(char *line){
```

Comment:
there is no need for a custom function here. there are several convenience functions for parsing and custom string processing in the utils namespace and you may also switch to use the Tokenizer/ValueTokenizer class for even more efficient and compact processing of parameter files.

Reply:
I removed this function.